### PR TITLE
Parse .cabal files; show error and warning diagnostics

### PR DESCRIPTION
**Pull request by runeksvendsen**
Depends on https://github.com/haskell/haskell-language-server/pull/2945

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2954"></a>

### CODE REVIEW

**Comment by runeksvendsen** [comment]
Wait... do all the plugins share the same dependency version? This plugin uses the parser from Cabal==3.6.2.0, and by looking at the CI logs it fails because it tries to use the same Cabal version for everything? :-/

**Comment by michaelpj** [comment]
All of the Haskell packages are compiled together into a single build plan by Cabal, yes. But I don't think any other plugins impose version constraints on Cabal. Possibly there's something to do with the Cabal library being bundled with GHC? It wouldn't shock me if we were only able to enable this on newer GHCs which give us a newer cabal library.

File: CODEOWNERS
Comment (kokobd) [inline CODEOWNERS:12]: Give @runeksvendsen write access?

**Comment by kokobd** [comment]
Mention it in the docs? For example section "Features", "Supported GHC Versions", etc.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (kokobd) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:35]: Can we simplify the warning message? I'm seeing
```
[parsing] [E] /home/kokobd/work/github.com/haskell/haskell-language-server/plugins/hls-cabal-plugin/hls-cabal-plugin.cabal:9:25:
unexpected Unknown SPDX license identifier: 'MIT2'
```
But I believe the file position is not needed, as the message already shows at that location.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs
Comment (hasufell) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs:43]:
```suggestion
    :: BS.ByteString -- ^ UTF-8 encoded bytestring
```

https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/src/Distribution.PackageDescription.Parsec.html#parseGenericPackageDescription shows that it actually checks for it... and does some pretty fishy stuff if it isn't

File: plugins/hls-cabal-plugin/hls-cabal-plugin.cabal
Comment (fendor) [reply plugins/hls-cabal-plugin/hls-cabal-plugin.cabal:34]: iirc, it would not be a problem to use the same cabal version for all GHC versions here, since Cabal 3.6 still compiles with GHC 8.8. However, stack makes it difficult, as we would have to either allow newer Cabal versions for all dependencies, or make the old resolver use the newer version for virtually all dependencies (e.g. packages that needed only version revisions). Don't know what do exactly do here, however, pre-built executables aren't build with stack so it should always use Cabal 3.6, afaict.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:35]: They are pretty much unstructured, I am not entirely sure that's possible :( but we can give it a try, nevertheless.

**Comment by pepeiborra** [comment]
@runeksvendsen are you able to wrap up this PR?

**Comment by fendor** [comment]
@pepeiborra I am making sure this PR will be pushed over the finish line. At the moment, we struggle with indeterministic test failure. Also, I want to add more tests, before merging.

**Comment by michaelpj** [comment]
What's the status of this now? It looks like there's a stack of dependent PRs, can we merge this as is and iterate?

**Comment by fendor** [comment]
@michaelpj We made great progress over MuniHac, I had hoped I fixed the CI on monday, I have to check what's going wrong.

**Comment by fendor** [comment]
@runeksvendsen are you interested in code ownership (in the sense of, you will be pinged if there are issues)?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:35]: Resolved for a follow-up PR, so this is no longer blocks the other hls-cabal-plugin PRs.

File: plugins/hls-cabal-plugin/hls-cabal-plugin.cabal
Comment (michaelpj) [reply plugins/hls-cabal-plugin/hls-cabal-plugin.cabal:36]: I think this generally isn't too much of a problem: this plugin mostly operates on cabal files, and different versions of cabal usually treat cabal files the same; differences in behaviour are controlled with the `cabal-version` field. So I think we're okay. BUT worth a comment :)

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs:45]: good comment

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs:59]: Could be worth a few simple unit tests for this function specifically.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:26]: use `OverloadedStrings`?

Comment (michaelpj) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:26]: I think this is not the most informative thing to put in the `source` field. Parsing what? I think this should probably just say "cabal".

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:41]: Should we have a version of this that works on LSP positions and put it with the other position utilities? Seems like a thing other people might want.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs:32]: why not inline this?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs:37]: Surely not! Won't this kill HLS??? We definitely need to forbid this function in hlint lol

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:112]: Nothing here seems to be cabal-specific, is this not a generic function we use elsewhere? Also why do we need it?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:235]: I get that we have a separate bit of the IDE state for tracking the cabal files of interest, but it seems like the logic for everything else is the same as for Haskell files of interest. Can we not generalize it? It sees like the only thing we'd need is to be generic over a `IO (Var (HashMap NormalizedFilePath FileOfInterestStatus))` or something?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:137]: Do we not have a function for this? Seems pretty standard: get the VFS contents or else read from disk.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:135]: I'm not sure what this log means. It doesn't seem like anything is being _modified_ here?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:148]: This logging seems very generic. Maybe we should put it in the shake code so it gets logged for every rule that produces diagnostics?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:157]: obscure :thinking:

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:163]: But it's not called by `restartCabalShakeSession` or indeed anywhere nearby. This whole comment is mysterious :thinking:

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:195]: I get that we do this for Haskell too... but I wonder why we can't e.g. just traverse the VFS and look for files matching the extension we care about?

File: plugins/hls-cabal-plugin/test/Main.hs
Comment (michaelpj) [inline plugins/hls-cabal-plugin/test/Main.hs:120]: I'd probably just shove the expected output in a file, but whatever.

**Review by michaelpj — APPROVED** [review]
Needs documentation!

**Comment by michaelpj** [comment]
i.e. something in `features.md`, plugin support status.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs:37]: LOL, good catch

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:163]: true it is obscure, but ghcide and hlint do the same thing. This part of the codebase in ghcide is barely documented but absolutely crucial afaict.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:157]: hlint and ghcide do the same thing, you can query for the `kick` function in ghcide.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:195]: having a separate variable seemed like a better seperation to me. Otherwise, you need to remember to always call the correct function...

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:163]: :cry:

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/LicenseSuggest.hs:59]: there are some integration tests, though. You are obviously right, adding two unit tests won't hurt at all, but let me be a horrible software engineer :angry:

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Parse.hs:37]: Code wasn't used :angel:

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:137]: Afaict, we do not, we have 
```haskell
moduleText :: (IsString e, MonadLsp c m) => Uri -> ExceptT e m Text
moduleText uri =
    handleMaybeM "mdlText" $
      (virtualFileText <$>)
          <$> getVirtualFile
              (toNormalizedUri uri)
 ```
 in hls-eval-plugin and 
 ```haskell
     provider :: forall m. FormattingMethod m => SMethod m -> PluginMethodHandler a m
    provider m ide _pid params
      | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri uri = do
        mf <- getVirtualFile $ toNormalizedUri uri
        case mf of
          Just vf -> do
            let typ = case m of
                  STextDocumentFormatting -> FormatText
                  STextDocumentRangeFormatting -> FormatRange (params ^. J.range)
                  _ -> error "mkFormattingHandlers: impossible"
            f ide typ (virtualFileText vf) nfp opts
          Nothing -> pure $ Left $ responseError $ T.pack $ "Formatter plugin: could not get file contents for " ++ show uri
```

in hls-plugin-api. Not in the mood for refactoring that right now.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:163]: what documentation would you like to see here? I am eager to share and discuss my finding about the internals of hls, but this is the best I can do in this situation

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Diagnostics.hs:44]: Added a function to hls-plugin-api

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs]: Not quite due to pattern matching, etc... I think we'd need a bit more infrastructure for that. Don't think that should be added here.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs]: Improved docs slightly.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:142]: The intent is to say, that we log the modification time of the file we want to parse. Afaict, this makes sure that our rule depends on the last modification time of that file, and only runs the rule, if the dependency is dirty.

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:144]: It looks like, most  Core IDE functions don't use the virtual files at all... very confusing...

Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:144]: ghcide uses `getFileContentsRule` and `getFileContents` to abstract over `virtualFileText`.

Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs]: It is a bit too complicated to re-use everything, but now it is better.

Comment (michaelpj) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs]: Do you want to quickly note down your findings in an issue or something so we can pick it up later if we want to?

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (michaelpj) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:142]: Okay, I kind of get that (but write it down!), I just find the log confusing. If it was `LogParsingCabalFile file t` that would make more sense to me.

Comment (michaelpj) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:142]: And now it arguably doesn't tell us much, given that we should be relying on GetFileContents?

**Review by michaelpj — APPROVED** [review]
LGTM!

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs:142]: I added a comment, but I am not sure we need more than that

File: plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs
Comment (fendor) [reply plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs]: In other words, we *do* have a function for it, a rule to be precise. The rule `GetFileContents` abstracts over reading file contents from the virtual file, but we have no rule that alternatively reads the file from disk if it cannot be found on the virtual disk. Not sure that requires an issue.

**Pull request merged**
